### DISCUSSION TITLE
Can you record "grey market" IPTV channels in SageTV?

### DISCUSSION THREAD

**Discussion by wayner9** [OpenSageTV/Community#154] [General]
I am getting very annoyed with my cable company and I am trying out an IPTV service.  This gives me access to thousands of channels and the channels are defined in an m3u file, I believe.  Is there any way to use SageTV to record the streaming video from these types of services?  I never actually used DCT much, but this seems like the kind of thing that it could do if you could feed it the correct stream.

**Comment by TGsell1** [comment]
I have a very similar question since my cable company threatened to disable my CableCard.   When I called them said they were backing off that so not now… but it’s enough of a threat.  So yes, what have Sage users and developers done to enable streaming DVR’ing?  I can do it manually through my  mini Pc boxes and network encoders so I just need a script that talks to the box and gets it to stream the show I want.   Manual might not be horrible.  Sent from my iPhoneOn Aug 28, 2024, at 5:59 PM, wayner9 ***@***.***> wrote:﻿
I am getting very annoyed with my cable company and I am trying out an IPTV service.  This gives me access to thousands of channels and the channels are defined in an m3u file, I believe.  Is there any way to use SageTV to record the streaming video from these types of services?  I never actually used DCT much, but this seems like the kind of thing that it could do if you could feed it the correct stream.

**Comment by mickgeller** [comment]
Yes, but it cannot handle the m3u. The guide data will drive Sagetv. I use Tvheadend for my IPTV source and Channels DVR for my streaming sources. Both handle M3U sources and export XMLTV guide data that Sagetv can consume. The export drives tuner definition in Sage and OpenDCT. Grab the best XML import plugin from the old forum. It's by hvymetal at "Updated XMLTV Plugin With Compile Step," first entry in the thread. No, you don't need to compile anything.  Find it in the EPG subforum. BTW, SageTV is still my primary DVR, becauee of my addiction to intelligent recording.

**Reply by wayner9** [reply]
Any pointers on more details on how you do all of this?  It looks like TVHeadend and Channels DVR are available as docker containers.  Is that how you run them?
For many of the channels I could just use guide data from Schedules Direct, as many of the channels are what I have today on cable.

**Reply by mickgeller** [reply]
Yes, I run both as Dockers. Tvheadend allow you to input the m3u, but you need to match the guide data, but that's a one-time thing. Frequently, I use the schedule direct data, not the guide data from the IPTV provider. I do this for channels that are on my local schedules direct guide. I'm in the US, but I also have a UK guide on my schedules direct account, so I can map data for those channels too. Out of the box, TV head in takes input from a number of sources including schedules Direct and XML TV. The latter is usually provided by the IPTV provider. Channels DVR is a bit easier, I think, and less work. It comes with plenty of documentation. TVheadend is something that I found I had to figure out on my own quite a bit.

**Reply by wayner9** [reply]
Thanks - do you edit your m3u down to something manageable?  Mine has tens of thousands of lines as it has all of the channels, plus all kinds of movies and TV series.

**Reply by mickgeller** [reply]
Yes, I edit it first.

**Reply by wayner9** [reply]
What's the best way to edit the m3u down.  The base one from my IPTV service, which includes movies and TV shows is 1/4 million lines long.

**Comment by TGsell1** [comment]
Does anyone have solutions for windows?Sent from my iPhoneOn Aug 28, 2024, at 11:08 PM, wayner9 ***@***.***> wrote:﻿
Thanks - do you edit your m3u down to something manageable?  Mine has tens of thousands of lines as it has all of the channels, plus all kinds of movies and TV series.

**Reply by wayner9** [reply]
Most of this stuff being discussed should work in Windows as well.  I am just figuring it all out myself, I am not sure what the function is all of these programs, and I have never used OpenDCT other than playing around with it a bit.  It appears that there is a URL that is the channel's stream.  For example this URL https://lnc-wkbw.tubi.video/playlist.m3u8 shows the channel WKBW.  You can copy this URL go to VLC, click on Open Network Stream and paste this URL.  You should then have that channel streaming.  So we just need a way to capture and record this stream.  And we need a way to get the guide data.

**Reply by mickgeller** [reply]
The issue, noted by TGsell1, is that TVHeadend is Linux only. I am not sure what Windows software can perform the functions of consume an M3U file, consume Guide data, allow matching of the two, and output XMLTV format for SageTV.  Regarding OpenDCT, it allows SageTV to pass it a channel number, so it knows what stream to start.

**Reply by wayner9** [reply]
How reliable is your recording via OpenDCT?  Is it flaky or is it pretty solid?  It seems like some recording software can have problems when ads get inserted and or if the stream buffers or has other hiccups?  I understand that this will be somewhat dependent on your provider, but would this be a viable way to record content rather than using cable?

**Comment by Ugadata** [comment]
In this forum, under Software/OpenDCT are links to threads in old forum for using OpenDCT - including links for using ChannelsDVR with OpenDCT.

I am not currently using ChannelsDVR as a source for Sage, But I did have it working awhile back. I still have ChannelsDVR running, just not using it with Sage.  

My Sage server is running Win 7.  SageTV, OpenDCT, and ChannelsDVR are all running on this server.  At the time I was playing with this setup, the hardware was an Intel NUC.  I have since moved the install to a virtual machine on ProxMox.

Keep in mind that TV Everywhere is how ChannelsDVR is able to access content on the streaming service.  Not all channels on the streaming services support  TVE (TV Everywhere).

**Reply by wayner9** [reply]
I am not using TV Everywhere as I am using a "grey" market IPTV service that gives me access to thousands of TV channels from around the world.

**Comment by nyplayer** [comment]
SageTV does not support M3U natively and is long in the tooth ... With Channels DVR I also was Feeding M3u to Opendct and getting TUBI , Friendly TV, TVE and Pluto in SageTV.  But I stopped paying for SCHEDDIRECT and Instead put the fee towards Channels DVR.

**Reply by wayner9** [reply]
Can you point me in the direction of where to learn more about how to do this.  I see folks talking about both Channels DVR and TVHeadend, but I am not sure exactly what I need to do this, and how to configure OpenDCT.

**Reply by nyplayer** [reply]
https://github.com/OpenSageTV/Community/discussions/51

**Reply by wayner9** [reply]
Thanks.  Now to figure out what that all means!  This seems like it could be super awesome if I get it all working, but could be a time sink until I get there.

**Reply by wayner9** [reply]
What role does Channels DVR play?  I thought that it was DVR software and was pretty much the same as SageTV.  Or is it just used to get streams from the IPTV service and forwards them on to OpenDCT?

**Reply by mickgeller** [reply]
Channels DVR can define a tuner via m3u. It also grabs guide data and provides an easy way to define that. Further, it can output the XMLTV guide data that SageTV can use.

**Reply by wayner9** [reply]
Thanks, how does TVHeadend fit into this as well.

**Reply by mickgeller** [reply]
TVHeadend can define a tuner via m3u. It also grabs guide data and provides an easy way to define that. Further, it can output the XMLTV guide data that SageTV can use. 
TVHeadend is less automated. With an M3U you end up potentially matching each entry manually for the guide data assignment. If you want to use Schedules Direct, it will, most likely, ensure the need for manual matching. There is additional work to even get to that point. 

To input the M3U, TVHeadend defines a "network". Each stream in the M3U defines a "MUX". The user triggers a scan to test each stream and, if good, creates a "Service" for each channel. The user then maps each service and that creates the "channel." At that point, the user checks the guide data to see if it is mapped, or manually map the channel to guide data correctly to the guide data. Channels with or without guide data can be in exported EPG data, and SageTV will create channels for them It is quite a bit of work, but it is one-time setup. Once it is set up, it requires minimal work to keep it going. TVheadend handles Schedules Direct very nicely as an EPG source. Note that TVheadend is Linux only.

**Comment by Ugadata** [comment]
ChannelsDVR could be used as a DVR by itself.  Whatever source is available to Channels can be made available to Sage through OpenDCT.
You can also get the guide data from Channels by using the updated XMLTV plugin.

TV Everywhere is how Channels is able to record many of the streaming channels.  Something similar to PlayOn.

**Comment by emveepee** [comment]
For multiple platform NextPVR supports IPTV from m3u and xsteam  but also check out Emby and Jellfyin    This discussion should probably not be about pirate IPTV especially as an alternative to cable since discussion  is banned on this and most public platforms   Even the open m3u files are being shutdown  so a lot of this is don't ask don't tell.

**Reply by wayner9** [reply]
NextPVR only has a client for Windows correct?  No client for set top boxes? Do NextPVR users use NextPVR as the back end to record and then use Plex, Emby or Jellyfin as the front end?

**Comment by Swagger200** [comment]
Not a direct equivalent, but I am able to record streaming channels from an Nvidia Shield Pro and a PC HDMI Out using a cheap Amazon HDMI to Component adapter, feeding a Hauppauge HD-PVR, controlled and recorded by SageTV. Quality is reasonably good.

**Reply by wayner9** [reply]
But you shouldn't have to do this with IPTV.  You can access streams via a URL so assuming that you can pipe the video stream to SageTV then you would be able to capture such streams without needing a hardware capture device like an HD-PVR.  And now worry about needing to tune channels.

OpenDCT can do stuff like this - I just have to figure it all out.

**Comment by TGsell1** [comment]
That’s good to hear.   How does Sage control the streaming show? Sent from my iPhoneOn Sep 4, 2024, at 5:41 PM, Swagger200 ***@***.***> wrote:﻿
Not a direct equivalent, but I am able to record streaming channels from an Nvidia Shield Pro and a PC HDMI Out using a cheap Amazon HDMI to Component adapter, feeding a Hauppauge HD-PVR, controlled and recorded by SageTV. Quality is reasonably good.

**Comment by Swagger200** [comment]
SageTV can't control the streaming box. I just record from the Component input on my Hauppauge HD-PVR, and setup a scheduled manual recording.

**Reply by Swagger200** [reply]
Or at least I haven't been able to get SageTV to control it. It may well be possible.

**Reply by wayner9** [reply]
IR control should work as emveepee indicated above, and in some instances you can get IP to work.

**Reply by Zogg44** [reply]
This is where you have to get creative. Depending on your host (Windows or Linux), you can use the EXEMultiTuner or gentuner plugin to use an external app to send IR commands through the HD-PVR blaster, or a USB-UIRT, or any other program you can find that will control that box. It will probably take some experimentation to find the right program or method to do it, then set up the plugin.

**Reply by wayner9** [reply]
With a USB-UIRT you don't need EXEMultituner or gentuner to send IR.  It is built into the core SageTV.  You just need the right IR codes and maybe a blaster pointed to the right spot of the box.
I am using gentuner for IP control, but I have a unique setup that let's me use IP control by using my Control4 Home Automation system to send the IP tuning commands.  Control4 seems to be the only system that has IP drivers for controlling these boxes. 
emveepee and I had a long discussion about this in another thread on this forum.

**Reply by Zogg44** [reply]
> With a USB-UIRT you don't need EXEMultituner or gentuner to send IR. It is built into the core SageTV. 

That's right. I was also using a unique setup with Unraid where I sent commands to a Windows box that then controlled the USB-UIRT, but if he can do it directly all the better.

**Comment by TGsell1** [comment]
Perfect Sent from my iPhoneOn Sep 5, 2024, at 9:52 AM, Swagger200 ***@***.***> wrote:﻿
SageTV can't control the streaming box. I just record from the Component input on my Hauppauge HD-PVR, and setup a scheduled manual recording.

**Comment by TGsell1** [comment]
I actually can record streaming directly from my mini PCs running windows MC on win 8.1    If there was a way to coordinate Sage manual recording and the streaming channel that would pretty much solve any issues with eliminating my CaeCard Sent from my iPhoneOn Sep 5, 2024, at 9:52 AM, Swagger200 ***@***.***> wrote:﻿
SageTV can't control the streaming box. I just record from the Component input on my Hauppauge HD-PVR, and setup a scheduled manual recording.

**Reply by wayner9** [reply]
What are you using to record the stream?  Are you using OpenDCT Generic HTTP capture functionality?  And are you capturing from something like an MPEG-TS stream defined in a URL?  For example, this is the URL for the Schwab network TV channel https://content.uplynk.com/channel/f9aafa1f132e40af9b9e7238bc18d128.m3u8 .  Could you capture this stream?  I am able to enter this URL in VLC and watch the channel.  Now how to figure out the rest of it.

**Comment by TGsell1** [comment]
I’m using the network encoder from URay.  It captures HDMI from the mini pc which is also controlled via a VNC so I can do everything remote from my phone … but manually.  Sent from my iPhoneOn Sep 5, 2024, at 4:19 PM, wayner9 ***@***.***> wrote:﻿
What are you using to record the stream?  Are you using OpenDCT Generic HTTP capture functionality?  And are you capturing from something like an MPEG-TS stream defined in a URL?  For example, this is the URL for the Schwab network TV channel https://content.uplynk.com/channel/f9aafa1f132e40af9b9e7238bc18d128.m3u8 .  Could you capture this stream?  I am able to enter this URL in VLC and watch the channel.  Now how to figure out the rest of it.

**Reply by wayner9** [reply]
Ok thanks.  So you are using a hardware encoder.  I am going to see if I can get this going with software only - it would be awesome if it works.

**Reply by wayner9** [reply]
aa

**Comment by TGsell1** [comment]
Agree! But I already have a great system going w WMC and Sage tuning via EXEMulti … this is my contingency plan for no CableCard.   Stay in touch on this! Sent from my iPhoneOn Sep 5, 2024, at 5:47 PM, wayner9 ***@***.***> wrote:﻿
Ok thanks.  So you are using a hardware encoder.  I am going to see if I can get this going with software only - it would be awesome if it works.

**Reply by wayner9** [reply]
Is it possible to communicate via email?  Can you think of any way to connect other than posting an email address here?  Github doesn't have private messages the way that the old forum did.

**Comment by TGsell1** [comment]
tomcgsell at the popular google mail
